jack-test/test/java/com/rapleaf/jack/transaction/TestDbMetrics.java

```diff
@@ -0,0 +1,200 @@
+
+    Future<Long> future1 = executorService.submit(() -> transactor.query(a -> {
+      sleepMillis(100);
+      return stopwatch.elapsedMillis();
+    }));
+    final Long[] startingTime2 = new Long[1];
+    Future future2 = executorService.submit(
+        () -> {
+          startingTime2[0] = stopwatch.elapsedMillis();
+          transactor.execute(a -> {sleepMillis(100);});
+        });
+    long finishingTime1 = future1.get();
+    future2.get();
+    DbMetrics dbMetrics = transactor.getDbMetrics();
+    double averageConnectionsWaitingTime = dbMetrics.getAverageConnectionWaitingTime();
+    transactor.close();
+    double expectedAverageConnectionsWaitingTime = (finishingTime1 - startingTime2[0]) / 2;
+    expectedAverageConnectionsWaitingTime = (expectedAverageConnectionsWaitingTime > 0) ? expectedAverageConnectionsWaitingTime : 0;
+    assertRoughEqual(averageConnectionsWaitingTime, expectedAverageConnectionsWaitingTime, 15);
+  }
+
+  @Test
+  public void testAverageConnectionExecutionTime() throws Exception {
+    TransactorImpl<IDatabase1> transactor = transactorBuilder.setMaxTotalConnections(2).get();
+    executorService = Executors.newFixedThreadPool(5);
+    Future<Long> future1 = executorService.submit(() -> {
+      long start = stopwatch.elapsedMillis();
+      transactor.execute(a -> {sleepMillis(100);});
+      return stopwatch.elapsedMillis() - start;
+    });
+    Future<Long> future2 = executorService.submit(() -> {
+      long start = stopwatch.elapsedMillis();
+      transactor.execute(a -> {sleepMillis(100);});
+      return stopwatch.elapsedMillis() - start;
+    });
+    double expectedAverageConnectionExecutionTime = ((double)future1.get() + (double)future2.get()) / 2;
+    DbMetrics dbMetrics = transactor.getDbMetrics();
+    double averageConnectionExecutionTime = dbMetrics.getAverageConnectionExecutionTime();
+    transactor.close();
+
+    assertRoughEqual(averageConnectionExecutionTime, expectedAverageConnectionExecutionTime, 20);
+  }
+
+  @Test
+  public void testAverageIdleConnections() throws Exception {
+    TransactorImpl<IDatabase1> transactor = transactorBuilder.setMaxTotalConnections(2).setMinIdleConnections(1).setKeepAliveTime(Duration.millis(50)).get();
+    transactor.execute(a -> {sleepMillis(100);});
+    long startIdleTime = stopwatch.elapsedMillis();
+    sleepMillis(100);
+    long idleTime = stopwatch.elapsedMillis() - startIdleTime;
+    DbMetrics dbMetrics = transactor.getDbMetrics();
+    long lifeTime = dbMetrics.getLifeTime();
+    double expectedAverageIdleConnections = (double)idleTime / (double)lifeTime;
+    double averageIdleConnectionsMaxValue = dbMetrics.getAverageIdleConnectionsMaxValue();
+    double averageIdleConnectionsMinValue = dbMetrics.getAverageIdleConnectionsMinValue();
+    transactor.close();
+
+    assert ((expectedAverageIdleConnections - .1 <= averageIdleConnectionsMaxValue) && (expectedAverageIdleConnections + .1 >= averageIdleConnectionsMinValue));
```

Comment:
Actually one of the comparisons is with averageIdleConnectionsMinValue and the other with averageIdleConnectionsMaxValue so I can't use assertRoughEqual